extensions/notebook/src/jupyter/jupyterServerInstallation.ts

```diff
@@ -356,6 +387,106 @@ export class JupyterServerInstallation {
+		installedPackages.forEach(pkg => pkgVersionMap.set(pkg.name, pkg.version));
+
+		let packagesToInstall: PythonPkgDetails[];
+		if (forceInstall) {
+			packagesToInstall = this._expectedPythonPackages;
+		} else {
+			packagesToInstall = [];
+			this._expectedPythonPackages.forEach(expectedPkg => {
+				let installedPkgVersion = pkgVersionMap.get(expectedPkg.name);
+				if (!installedPkgVersion || utils.comparePackageVersions(installedPkgVersion, expectedPkg.version) < 0) {
+					packagesToInstall.push(expectedPkg);
+				}
+			});
+		}
+
+		if (packagesToInstall.length > 0) {
+			let doUpgrade: boolean;
+			if (promptForUpgrade) {
+				doUpgrade = await this._prompter.promptSingle<boolean>(<IQuestion>{
+					type: QuestionTypes.confirm,
+					message: localize('confirmPipUpgrade', "Some installed pip packages need to be upgraded. Would you like to upgrade them now?"),
+					default: true
+				});
+			} else {
+				doUpgrade = true;
+			}
+			if (doUpgrade) {
+				await this.installPipPackages(packagesToInstall, true);
+			}
+		}
+	}
+
+	private async upgradeCondaPackages(promptForUpgrade: boolean, forceInstall: boolean): Promise<void> {
+		let condaPackagesToInstall: PythonPkgDetails[] = [];
+		let pipPackagesToInstall: PythonPkgDetails[] = [];
+
+		if (forceInstall) {
+			condaPackagesToInstall = this._expectedCondaPackages;
+			pipPackagesToInstall = this._expectedCondaPipPackages;
+		} else {
+			condaPackagesToInstall = [];
+			pipPackagesToInstall = [];
+
+			// Conda packages
+			let installedCondaPackages = await this.getInstalledCondaPackages();
+			let condaVersionMap = new Map<string, string>();
+			installedCondaPackages.forEach(pkg => condaVersionMap.set(pkg.name, pkg.version));
+
+			this._expectedCondaPackages.forEach(expectedPkg => {
+				let installedPkgVersion = condaVersionMap.get(expectedPkg.name);
+				if (!installedPkgVersion || utils.comparePackageVersions(installedPkgVersion, expectedPkg.version) < 0) {
+					condaPackagesToInstall.push(expectedPkg);
+				}
+			});
+
+			// Pip packages
+			let installedPipPackages = await this.getInstalledPipPackages();
+			let pipVersionMap = new Map<string, string>();
+			installedPipPackages.forEach(pkg => pipVersionMap.set(pkg.name, pkg.version));
+
+			this._expectedCondaPipPackages.forEach(expectedPkg => {
+				let installedPkgVersion = pipVersionMap.get(expectedPkg.name);
+				if (!installedPkgVersion || utils.comparePackageVersions(installedPkgVersion, expectedPkg.version) < 0) {
+					pipPackagesToInstall.push(expectedPkg);
+				}
+			});
+		}
+
+		if (condaPackagesToInstall.length > 0 || pipPackagesToInstall.length > 0) {
+			let doUpgrade: boolean;
+			if (promptForUpgrade) {
+				doUpgrade = await this._prompter.promptSingle<boolean>(<IQuestion>{
+					type: QuestionTypes.confirm,
+					message: localize('confirmCondaUpgrade', "Some installed conda and pip packages need to be upgraded. Would you like to upgrade them now?"),
+					default: true
```

Comment:
Might be good to indicate what packages are being updated with what from/to version

Reply:
The current prompt framework only takes a single line of input, so there's not enough space to append all the package info.